### ISSUE TITLE
ProvisionedThroughputExceededException not automatically retried for putItem requests

### ISSUE THREAD

**Issue by mikemherron** [aws/aws-sdk-php#815]
It looks like ProvisionedThroughputExceededException are not being correctly retried by the SDK when invoking putItem and exceeding a tables provisioned throughput. Adding debug print statements to RetryMiddleware.php suggests that the delay is being calculated but not applied.

The below test script, placed in the root of a directory where the SDK is installed via composer, can be used to demonstrate this:

```php
<?php

include "vendor/autoload.php";

use Aws\DynamoDb\DynamoDbClient;

/*
 Add the below line of code to vendor/aws/aws-sdk-php/src/RetryMiddleware.php, around 
 line 120 to print out the retry delays that should be applied:
 
 <code>
 // Delay fn is called with 0, 1, ... so increment after the call.
 $command['@http']['delay'] = $delay($retries++);
 + print gmdate("H:i:s") . ":Retrying with delay:" . $command['@http']['delay'] . "\n";
 return $handler($command, $request)->then($g, $g);
 </code>
*/

function out( $msg ) {
    print gmdate("H:i:s") . ": " . $msg . "\n";
}

$options = [
    'region'    => "eu-west-1",
    'endpoint'  => "https://dynamodb.eu-west-1.amazonaws.com",
    'credentials' => [
        'key'       => "",
        'secret'    => ""
    ],
    'version'   => 'latest',
];

$client = new DynamoDbClient($options);

$tables = $client->listTables();

if ( !in_array( 'TestThroughputExceptions', $tables['TableNames' ] ) ) {   

    out( "Creating TestThroughputExceptions table");

    $client->createTable( [
        'TableName' => 'TestThroughputExceptions',
        'AttributeDefinitions' => [ 
            [ 'AttributeName' => 'ID',
              'AttributeType' => 'S' ] 
        ],
        'KeySchema' => [ 
            [ 'AttributeName' => 'ID',
              'KeyType' => 'HASH' ] 
        ],
        'ProvisionedThroughput' => [
            'ReadCapacityUnits'  => 1,
            'WriteCapacityUnits' => 1
        ]
    ]);
    
    $client->waitUntil( 'TableExists', [
        'TableName' => 'TestThroughputExceptions'
    ]);
}


while( true ) {
    out( "Starting putItem");
    $client->putItem( [
        'TableName'     => 'TestThroughputExceptions',
        'Item'          => [ 'ID' => ['S' => 'SomeID' ] ]
    ]);
    out( "Finished putItem");
}
```

Running 2 instances of this script at the same time produces the below output:

```
17:22:29: Starting putItem
17:22:29: Finished putItem
17:22:29: Starting putItem
17:22:29: Finished putItem
17:22:29: Starting putItem
17:22:29: Finished putItem
17:22:29: Starting putItem
17:22:29:Retrying with delay:0
17:22:29:Retrying with delay:0.05
17:22:29:Retrying with delay:0.1
17:22:29:Retrying with delay:0.2
17:22:29:Retrying with delay:0.4
17:22:30:Retrying with delay:0.8
17:22:30:Retrying with delay:1.6
17:22:30:Retrying with delay:3.2
17:22:30:Retrying with delay:6.4
17:22:30:Retrying with delay:12.8
17:22:30:Retrying with delay:25.6

Fatal error: Uncaught exception 'Aws\DynamoDb\Exception\DynamoDbException' with message 'Error executing "PutItem" on "https://dynamodb.eu-west-1.amazonaws.com"; AWS HTTP error: Client error response [url] https://dynamodb.eu-west-1.amazonaws.com [status code] 400 [reason phrase] Bad Request ProvisionedThroughputExceededException (client): The level of configured provisioned throughput for the table was exceeded. Consider increasing your provisioning level with the UpdateTable API - {"__type":"com.amazonaws.dynamodb.v20120810#ProvisionedThroughputExceededException","message":"The level of configured provisioned throughput for the table was exceeded. Consider increasing your provisioning level with the UpdateTable API"}' in /work/ChilliWorksConnectLaravel/vendor/aws/aws-sdk-php/src/WrappedHttpHandler.php on line 152
```

Looking at the log timestamps, there is no delay between the actual putItem request - assuming that the intended delays are in seconds rather than milliseconds.

**Comment by mikemherron** [comment]
After testing this a bit more, I think the issue may be that the delay is being implemented as milliseconds rather than seconds. Usually the output of the test script is as above (once the script starts 'Retrying', it never manages to successfully submit a request again), but have now seen the below happen:

```
17:50:56: Finished putItem
17:50:56: Starting putItem
17:50:56:Retrying with delay:0
17:50:56:Retrying with delay:0.05
17:50:56:Retrying with delay:0.1
17:50:56:Retrying with delay:0.2
17:50:56:Retrying with delay:0.4
17:50:56:Retrying with delay:0.8
17:50:56:Retrying with delay:1.6
17:50:56:Retrying with delay:3.2
17:50:56:Retrying with delay:6.4
17:50:56: Finished putItem
17:50:56: Starting putItem
17:50:56:Retrying with delay:0
17:50:56:Retrying with delay:0.05
17:50:56:Retrying with delay:0.1
17:50:56:Retrying with delay:0.2
17:50:57:Retrying with delay:0.4
17:50:57:Retrying with delay:0.8
17:50:57:Retrying with delay:1.6
17:50:57:Retrying with delay:3.2
17:50:57:Retrying with delay:6.4
17:50:57:Retrying with delay:12.8
17:50:57:Retrying with delay:25.6

```

So in this one case, it did manage to recover. but just failed again on the next request.

**Comment by mikemherron** [comment]
I think the delay is being passed to Guzzle in the GuzzleHandler just as it's defined in the RetryMiddleware, but Guzzle actually interprets this as as milliseconds. Hacking in the below change to my local vendor folder seems to resolve the issue (around line 100)

```
if (isset($options['delay'])) {
-        $ringConfig['delay'] = $options['delay'];
+       $ringConfig['delay'] = $options['delay'] * 1000;
            unset($options['delay']);
```

When running the 2 scripts side by side, you can then see the delays being implemented correctly.

However, not sure if this is the correct place for the change?

**Comment by jeskew** [comment]
It does look like [Guzzle is interpreting the value as milliseconds](http://guzzle.readthedocs.org/en/latest/request-options.html#delay). Thanks for the detailed report and all the investigation into this!

**Comment by mikemherron** [comment]
No problem.

Based on the comment in Waiter (" // Set the delay. (Note: handlers except delay in milliseconds."), feels like the fix should actually go in DynamoDB client where the delay function is provided to the RetryMiddleware?

```
$list->appendSign(
            Middleware::retry(
                RetryMiddleware::createDefaultDecider($value),
                function ($retries) {
                    return $retries
                        ? (50 * (int) pow(2, $retries - 1)) / 1000
                        : 0;
                }
            ),
            'retry'
        );
```
Remove the division by 1000?

**Comment by jeskew** [comment]
This was fixed in #816 and will go out in the next release.
